torch/csrc/autograd/engine.cpp

```diff
@@ -421,6 +443,18 @@ auto Engine::evaluate_function(FunctionTask& task) -> void {
           continue;
         }
       }
+
+      #ifdef USE_CUDA
+        // Switches to and syncs with child's stream (if needed)
+        const auto child_device = output.is_cuda() ? output.get_device() : -1;
+        auto child_stream = next.function->input_metadata(next.input_nr).stream();
+        auto_gpu_stream.set(child_device, child_stream);
```

Comment:
Do we really need to switch to the `child_stream` only to synchronize it with `cuda_event`?

Reply:
Also, it would be nicer to have an outer guard on `if (parent_device != -1)` just to make sure we don't waste any time looking up the metadata, and checking the `is_cuda()` if we already know that the function we're processing is a CPU one

Reply:
We're not switching to the child stream to synchronize the event, we're switching because we're about to put something in its input buffer, which we have to do on the device + stream for that input buffer slot. 

For the same reason we can't guard on the parent_device being -1 before this. We could add an outer guard on the child_device after we acquired its device.

Remember that this call is replacing the former DeviceGuard (nee AutoGPU) in InputBuffer which performed the same check.

Reply:
Ok, makes sense.